cmd/atlas/internal/cmdapi/migrate.go

```diff
@@ -297,40 +297,77 @@ func CmdMigrateApplyRun(cmd *cobra.Command, args []string) error {
 			}
 		}
 	}(rrw.(*entmigrate.EntRevisions), cmd.Context())
+	// Determine pending files and lock the database while working.
+	ex, err := migrate.NewExecutor(c.Driver, dir, rrw, migrate.WithLogger(l))
+	if err != nil {
+		return err
+	}
+	unlock, err := ex.Lock(cmd.Context())
+	if err != nil {
+		return err
+	}
+	defer unlock()
+	pending, err := ex.Pending(context.Background())
+	if err != nil && !errors.Is(err, migrate.ErrNoPendingFiles) {
+		return err
+	}
+	if errors.Is(err, migrate.ErrNoPendingFiles) {
+		cmd.Println("The migration directory is synced with the database, no migration files to execute")
+		return nil
+	}
+	// Cannot apply more than len(pending) files.
+	if n > 0 {
+		if n >= len(pending) {
+			n = len(pending)
+		}
+		pending = pending[:n]
+	}
+	revs, err := rrw.ReadRevisions(cmd.Context())
+	if err != nil {
+		return err
+	}
+	if err := migrate.LogIntro(struct {
+		migrate.Logger
+		migrate.Scanner
+	}{l, dir.(migrate.Scanner)}, revs, pending); err != nil {
+		return err
+	}
 	var (
 		drv migrate.Driver
 		tx  *sqlclient.TxClient
 	)
 	if MigrateFlags.DryRun {
 		drv = &dryRunDriver{c.Driver}
 		rrw = &dryRunRevisions{rrw}
-	} else {
-		// Wrap the whole execution in one transaction. This behaviour will change once
-		// there are insights about migration files available.
-		tx, err = c.Tx(cmd.Context(), nil)
+	}
+	for _, f := range pending {
+		if !MigrateFlags.DryRun {
+			// Wrap the file execution in one transaction.
+			tx, err = c.Tx(cmd.Context(), nil)
+			if err != nil {
+				return err
+			}
+			drv = tx.Driver
```

Comment:
What happens in the next iteration when calling `c.Tx()`? Also, we can consider using checkpoints in the future.

Reply:
Ignore this.